### PR TITLE
Add bytes API

### LINKED ISSUE #1196: Document dask.bytes core api
At least needs api docs, but some prose on the use/design may also be nice. A caveat should also be added that most users probably want the collection-specific functions (e.g. `dd.read_csv`, `db.read_text`, etc...).

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #1196

### CODE REVIEW

File: docs/source/bytes.rst [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+are quite particular and receive custom treatment.
+
+
+Delimiters
+----------
+
+The ``read_bytes`` function takes a path (or globstring of paths) and produces
+a sample of the first file and a list of delayed objects for each of the other
+files.  If passed a delimiter such as ``delimiter=b'\n'`` it will ensure that
+the blocks of bytes start and end on this delimiter.  This allows other
```
Comment (jcrist) [inline docs/source/bytes.rst]: May need to clarify for inclusivity. Currently this reads to me as returning "\nsometext\n".

**Comment by jcrist** [comment]
Besides my comments above, this looks good to me.

File: docs/source/bytes.rst (hunk 1 above)
Comment (mrocklin) [inline docs/source/bytes.rst]: Fixed

File: docs/source/bytes.rst [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   open_files
+   open_text_files
+
+These functions are extensible in their output formats (bytes, file objects),
+their input locations (file system, S3, HDFS), line delimiters, and compression
+
+These functions provide data as ``dask.delayed`` objects.  These objects either
+point to blocks of bytes (``read_bytes``) or open file objects (``open_files``,
+``open_text_files``).  They can handle different compressions by prepending
+protocols like ``s3://`` or ``hdfs://``.  They handle compressions listed in the
```
Comment (mrocklin) [inline docs/source/bytes.rst]: Fixed

File: docs/source/bytes.rst [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+.. currentmodule:: dask.bytes
+
+.. autosummary::
+   read_bytes
+   open_files
+   open_text_files
+
+These functions are extensible in their output formats (bytes, file objects),
+their input locations (file system, S3, HDFS), line delimiters, and compression
```
Comment (mrocklin) [inline docs/source/bytes.rst]: Fixed

**Comment by mrocklin** [comment]
Merging soon if no comments

**Pull request merged**